src/common/replication_common.cpp

```diff
@@ -569,6 +498,118 @@ void replica_helper::load_meta_servers(/*out*/ std::vector<dsn::rpc_address> &se
+                                          /*out*/ std::vector<std::string> &data_dir_tags,
+                                          /*out*/ std::string &err_msg)
+{
+    // - if {config_dirs_str} is empty (return true):
+    //   - dir = {default_dir}
+    //   - dir_tag/data_dir_tag = "default"
+    //   - data_dir = {default_dir}/"reps"
+    // - else if {config_dirs_str} = "tag1:dir1,tag2:dir2:tag3:dir3" (return true):
+    //   - dir1 = "dir1"/{app_name}
+    //   - dir_tag1/data_dir_tag1 = "tag1"
+    //   - data_dir1 = "dir1"/{app_name}/"reps"
+    // - else (return false):
+    //   - invalid format and set {err_msg}
+    std::vector<std::string> dirs;
+    std::vector<std::string> dir_tags;
+    utils::split_args(config_dirs_str.c_str(), dirs, ',');
+    if (dirs.empty()) {
+        dirs.push_back(default_dir);
+        dir_tags.push_back("default");
+    } else {
+        for (auto &dir : dirs) {
+            std::vector<std::string> tag_and_dir;
+            utils::split_args(dir.c_str(), tag_and_dir, ':');
+            if (tag_and_dir.size() != 2) {
+                err_msg = fmt::format("invalid data_dir item({}) in config", dir);
+                return false;
+            }
+            if (tag_and_dir[0].empty() || tag_and_dir[1].empty()) {
+                err_msg = fmt::format("invalid data_dir item({}) in config", dir);
+                return false;
+            }
+            dir = utils::filesystem::path_combine(tag_and_dir[1], app_name);
+            for (unsigned i = 0; i < dir_tags.size(); ++i) {
+                if (dirs[i] == dir) {
+                    err_msg = fmt::format("dir({}) and dir({}) conflict", dirs[i], dir);
+                    return false;
+                }
+            }
+            for (unsigned i = 0; i < dir_tags.size(); ++i) {
+                if (dir_tags[i] == tag_and_dir[0]) {
+                    err_msg = fmt::format(
+                        "dir({}) and dir({}) have same tag({})", dirs[i], dir, tag_and_dir[0]);
+                    return false;
+                }
+            }
+            dir_tags.push_back(tag_and_dir[0]);
+        }
+    }
+
+    for (unsigned i = 0; i < dirs.size(); ++i) {
+        std::string &dir = dirs[i];
+        ddebug_f("data_dirs[{}] = {}, tag = {}", i + 1, dir, dir_tags[i]);
+        data_dirs.push_back(utils::filesystem::path_combine(dir, "reps"));
+        data_dir_tags.push_back(dir_tags[i]);
+    }
+    return true;
+}
+
+/*static*/ void
+replication_options::get_data_dirs_in_black_list(const std::string &fname,
+                                                 /*out*/ std::vector<std::string> &dirs)
+{
+    if (fname.empty() || (!fname.empty() && !utils::filesystem::file_exists(fname))) {
+        ddebug_f("data_dirs_black_list_file[{}] not found, ignore it", fname);
+        return;
+    }
+
+    ddebug_f("data_dirs_black_list_file[{}] found, apply it", fname.c_str());
```

Comment:
ddebug_f("data_dirs_black_list_file[{}] found, apply it", fname);

Reply:
This pr is for refactor, so I copy old code into the new functions. I will update it.